repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DomainPreProcessor.java

```diff
@@ -0,0 +1,161 @@
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processCreateDomain");
+        AtlasObjectId parentDomainObject = (AtlasObjectId) entity.getRelationshipAttribute(DOMAIN_DOMAIN_PARENT_REL_TYPE);
+
+        if (parentDomainObject == null) {
+            entity.removeAttribute(PARENT_DOMAIN_QN);
+            entity.removeAttribute(SUPER_DOMAIN_QN);
+        } else {
+            AtlasVertex parentDomain = retrieverNoRelation.getEntityVertex(parentDomainObject);
+            String parentDomainQualifiedName = parentDomain.getProperty(QUALIFIED_NAME, String.class);
+
+            entity.setAttribute(PARENT_DOMAIN_QN, parentDomainQualifiedName);
+
+            String[] splitted = parentDomainQualifiedName.split("/");
+            String superDomainQualifiedName = String.format("%s/%s/%s", splitted[0], splitted[1], splitted[2]);
+            entity.setAttribute(SUPER_DOMAIN_QN, superDomainQualifiedName);
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private void processUpdateDomain(AtlasEntity entity, AtlasVertex vertex, EntityMutationContext context) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processUpdateDomain");
+
+        AtlasEntity storedDomain = entityRetriever.toAtlasEntity(vertex);
+        AtlasObjectId currentParent = (AtlasObjectId) storedDomain.getRelationshipAttribute(DOMAIN_DOMAIN_PARENT_REL_TYPE);
+
+        AtlasEntityHeader newParent = getNewParentDomain(entity, context);
+
+        if (currentParent == null && newParent != null) {
+            // attaching parent for the first time, set attributes as well
+
+            String parentDomainQualifiedName = (String) newParent.getAttribute(QUALIFIED_NAME);
+            entity.setAttribute(PARENT_DOMAIN_QN, parentDomainQualifiedName);
+
+            String superDomainQualifiedName = (String) newParent.getAttribute(SUPER_DOMAIN_QN);
+            if (StringUtils.isEmpty(superDomainQualifiedName)) {
+                String[] splitted = parentDomainQualifiedName.split("/");
+                superDomainQualifiedName = String.format("%s/%s/%s", splitted[0], splitted[1], splitted[2]);
+            }
+
+            entity.setAttribute(SUPER_DOMAIN_QN, superDomainQualifiedName);
+
+        } else {
+            //never update these attributes with API request
+            entity.removeAttribute(PARENT_DOMAIN_QN);
+            entity.removeAttribute(SUPER_DOMAIN_QN);
+            if (entity.getRelationshipAttributes() != null) {
+                entity.getRelationshipAttributes().remove(DOMAIN_DOMAIN_PARENT_REL_TYPE);
+            }
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private AtlasEntityHeader getNewParentDomain(AtlasEntity entity, EntityMutationContext context) throws AtlasBaseException {
```

Comment:
Can you please rename this function to setParent as we are using this name in other features.